hoomd/PythonLocalDataAccess.h

```diff
@@ -236,108 +230,120 @@ template<class Output, class Data> class LocalDataAccess
-     *  get_array_func. It is templated off of T (which means that if
-     *  U=GlobalArray then the full type is GlobalArray<T>)
-     *
-     *  Arguments:
-     *  handle: a reference to the unique_ptr that holds the ArrayHandle.
-     *  get_array_func: the method of m_data to use to access the array.
-     *  of the exposed array in Python.
-     *  read_only: whether the array should be read only (defaults to True).
-     */
-    template<class T, template<class> class U = GlobalArray>
-    Output getGlobalBuffer(std::unique_ptr<ArrayHandle<T>>& handle,
-                           const U<T>& (Data::*get_array_func)() const,
-                           bool read_only = true)
+    // clear should remove any references to ArrayHandle objects so the
+    // handle can be released for other objects.
+    virtual void clear() = 0;
+
+    /// Ensure that arrays are not accessed outside context manager.
+    inline void checkManager()
         {
-        checkManager();
-        updateHandle(handle, get_array_func, read_only);
+        if (!m_in_manager)
+            {
+            throw std::runtime_error("Cannot access arrays outside context manager.");
+            }
+        }
+
+    /// Helper function to acquire array handle if not already acquired.
+    template<class T, template<class> class U>
+    void updateHandle(std::unique_ptr<ArrayHandle<T>>& handle,
+                      const U<T>& (Data::*get_array_func)() const,
+                      bool read_only)
+        {
+        if (!handle)
+            {
+            auto mode = read_only ? access_mode::read : access_mode::readwrite;
+            handle = std::move(std::unique_ptr<ArrayHandle<T>>(
+                new ArrayHandle<T>((m_data.*get_array_func)(), Output::device, mode)));
+            }
+        }
+
+    /// object to access array data from
+    Data& m_data;
+    /// flag for being inside Python context manager
+    bool m_in_manager;
+    };
 
-        auto size = m_data.getNGlobal();
-        unsigned int* data = handle.get()->data;
+enum class GhostDataFlag
+    {
+    standard,
+    ghost,
+    both
+    };
 
-        return Output::make(data,
-                            std::vector<ssize_t>({size}),
-                            std::vector<ssize_t>({sizeof(T)}),
-                            true);
+///
+/** @brief Base class for accessing per-* Global or GPU arrays/vectors in Python.
+ *
+ *  Template Parameters:
+ *  Output - the output buffer class for the class should be HOOMDDeviceBuffer
+ *  or HOOMDHostBuffer
+ *  Data - the class of the object we wish to expose data from
+ *
+ *  The main methods of LocalDataAccess are getLocalBuffer and getGlobalBuffer
+ *  which provide a way to automatically convert an Global/GPUArray into an
+ *  object of type Output of a MPI local, MPI local with ghosts, or global size
+ *  (for particles bonds, etc.).
+ */
+template<class Output, class Data>
+class PerGroupLocalDataAccess : public LocalDataAccess<Output, Data>
```

Comment:
We use the word `group` to refer to bonded groups, so `PerGroup` sounds like it applies only to those classes. This class provides helpers for arrays with ghost data, so name it `GhostLocalDataAccess`.

Reply:
I'll rename it, but  this class also does per-particle buffers regardless of ghosts too.